File: gym_management/befit_gym_manager/report/revenue_report/revenue_report.js
// Copyright (c) 2025, SteveNgash and contributors
// For license information, please see license.txt

// frappe.query_reports["Revenue Report"] = {
// 	"filters": [

// 	]
// };
frappe.query_reports["Revenue Report by Month"] = {
    "filters": [
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "reqd": 0
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "reqd": 0
        }
    ],

    onload: function(report) {
        report.page.add_inner_button(__("Toggle Chart"), function() {
            frappe.query_report.toggle_chart();
        });

        report.page.add_inner_button(__("Refresh Report"), function() {
            frappe.query_report.refresh();
        });
    },

    "get_chart_data": function(columns, result) {
        if (!result || result.length === 0) {
            return { message: "No revenue data available for the selected date range." };
        }

        let labels = result.map(row => row.month);
        let values = result.map(row => row.total_revenue);

        return {
            data: {
                labels: labels,
                datasets: [
                    {
                        name: "Total Revenue (Ksh)",
                        values: values
                    }
                ]
            },
            type: "bar"
        };
    }
};

File: gym_management/befit_gym_manager/report/revenue_report/revenue_report.py
# Copyright (c) 2025, SteveNgash and contributors
# For license information, please see license.txt

# import frappe


# def execute(filters=None):
# 	columns, data = [], []
# 	return columns, data
import frappe
from frappe.utils import getdate

def execute(filters=None):
    filters = frappe._dict(filters or {})

    # Define report columns
    columns = get_columns()

    # Fetch data
    data = get_data(filters)

    # Generate chart data
    chart_data = get_chart_data(data)

    return columns, data, None, chart_data

def get_columns():
    """Define columns for the report."""
    return [
        {
            "label": "Month",
            "fieldname": "month",
            "fieldtype": "Data",
            "width": 120
        },
        {
            "label": "Total Revenue (Ksh)",
            "fieldname": "total_revenue",
            "fieldtype": "Currency",
            "width": 150
        }
    ]

def get_data(filters):
    """Fetch and process data for the report."""
    conditions = []

    if filters.get("from_date"):
        conditions.append(f"date(date) >= '{filters.from_date}'")

    if filters.get("to_date"):
        conditions.append(f"date(date) <= '{filters.to_date}'")

    where_clause = " AND ".join(conditions) if conditions else "1=1"

    query = f"""
        SELECT 
            DATE_FORMAT(date, '%Y-%m') AS month,
            SUM(revenue) AS total_revenue
        FROM 
            `tabGym Revenue`
        WHERE 
            {where_clause}
        GROUP BY 
            month
        ORDER BY 
            month ASC
    """

    return frappe.db.sql(query, as_dict=True)

def get_chart_data(data):
    """Prepare chart data for the report."""
    if not data:
        return None

    labels = [row["month"] for row in data]
    values = [row["total_revenue"] for row in data]

    return {
        "data": {
            "labels": labels,
            "datasets": [
                {
                    "name": "Total Revenue (Ksh)",
                    "values": values
                }
            ]
        },
        "type": "bar"
    }
